// opencc-js@1.3.1/dist/esm-lib/preset/cn2t.js
import from_cn from "../from/cn.js";
import to_hk from "../to/hk.js";
import to_tw from "../to/tw.js";
import to_twp from "../to/twp.js";
import to_jp from "../to/jp.js";
import dict_HKVariants from "../dict/HKVariants.js";
import dict_STCharacters from "../dict/STCharacters.js";
import dict_STPhrases from "../dict/STPhrases.js";
import dict_TWPhrases from "../dict/TWPhrases.js";
import dict_TWVariants from "../dict/TWVariants.js";

const fromDicts = {
    cn: from_cn
};

const toDicts = {
    hk: to_hk,
    tw: to_tw,
    twp: to_twp,
    jp: to_jp
};

const configs = {
    s2hk: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters], [dict_HKVariants]] },
    s2t: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters]] },
    s2tw: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters], [dict_TWVariants]] },
    s2twp: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters], [dict_TWPhrases], [dict_TWVariants]] }
};

export {fromDicts as from, toDicts as to, configs};
