// opencc-js@1.3.1/dist/esm-lib/preset/full.js
import from_cn from "../from/cn.js";
import from_hk from "../from/hk.js";
import from_tw from "../from/tw.js";
import from_twp from "../from/twp.js";
import from_jp from "../from/jp.js";
import to_cn from "../to/cn.js";
import to_hk from "../to/hk.js";
import to_tw from "../to/tw.js";
import to_twp from "../to/twp.js";
import to_jp from "../to/jp.js";
import dict_HKVariants from "../dict/HKVariants.js";
import dict_HKVariantsRev from "../dict/HKVariantsRev.js";
import dict_HKVariantsRevPhrases from "../dict/HKVariantsRevPhrases.js";
import dict_JPShinjitaiCharacters from "../dict/JPShinjitaiCharacters.js";
import dict_JPShinjitaiPhrases from "../dict/JPShinjitaiPhrases.js";
import dict_JPVariants from "../dict/JPVariants.js";
import dict_JPVariantsRev from "../dict/JPVariantsRev.js";
import dict_STCharacters from "../dict/STCharacters.js";
import dict_STPhrases from "../dict/STPhrases.js";
import dict_TSCharacters from "../dict/TSCharacters.js";
import dict_TSPhrases from "../dict/TSPhrases.js";
import dict_TWPhrases from "../dict/TWPhrases.js";
import dict_TWPhrasesRev from "../dict/TWPhrasesRev.js";
import dict_TWVariants from "../dict/TWVariants.js";
import dict_TWVariantsRev from "../dict/TWVariantsRev.js";
import dict_TWVariantsRevPhrases from "../dict/TWVariantsRevPhrases.js";

const fromDicts = {
    cn: from_cn,
    hk: from_hk,
    tw: from_tw,
    twp: from_twp,
    jp: from_jp
};

const toDicts = {
    cn: to_cn,
    hk: to_hk,
    tw: to_tw,
    twp: to_twp,
    jp: to_jp
};

const configs = {
    hk2s: { segmentation: dict_TSPhrases, conversionChain: [[dict_HKVariantsRevPhrases, dict_HKVariantsRev], [dict_TSPhrases, dict_TSCharacters]] },
    hk2t: { segmentation: dict_HKVariantsRevPhrases, conversionChain: [[dict_HKVariantsRevPhrases, dict_HKVariantsRev]] },
    jp2t: { segmentation: dict_JPShinjitaiPhrases, conversionChain: [[dict_JPShinjitaiPhrases, dict_JPShinjitaiCharacters, dict_JPVariantsRev]] },
    s2hk: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters], [dict_HKVariants]] },
    s2t: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters]] },
    s2tw: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters], [dict_TWVariants]] },
    s2twp: { segmentation: dict_STPhrases, conversionChain: [[dict_STPhrases, dict_STCharacters], [dict_TWPhrases], [dict_TWVariants]] },
    t2hk: { segmentation: dict_HKVariants, conversionChain: [[dict_HKVariants]] },
    t2jp: { segmentation: dict_JPVariants, conversionChain: [[dict_JPVariants]] },
    t2s: { segmentation: dict_TSPhrases, conversionChain: [[dict_TSPhrases, dict_TSCharacters]] },
    t2tw: { segmentation: dict_TWVariants, conversionChain: [[dict_TWVariants]] },
    tw2s: { segmentation: dict_TSPhrases, conversionChain: [[dict_TWVariantsRevPhrases, dict_TWVariantsRev], [dict_TSPhrases, dict_TSCharacters]] },
    tw2sp: { segmentation: dict_TSPhrases, conversionChain: [[dict_TWPhrasesRev, dict_TWVariantsRevPhrases, dict_TWVariantsRev], [dict_TSPhrases, dict_TSCharacters]] },
    tw2t: { segmentation: dict_TWVariantsRevPhrases, conversionChain: [[dict_TWVariantsRevPhrases, dict_TWVariantsRev]] }
};

export {fromDicts as from, toDicts as to, configs};
